File: test/load.test.ts
import { mkdirSync, mkdtempSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { afterEach, describe, expect, it, vi } from "vitest";
import { defaults, SCHEMA_URL } from "../src/config";
import { FILE_NAME, globalPath, load } from "../src/load";

function tempDir(): string {
  return mkdtempSync(path.join(tmpdir(), "commitron-test-"));
}

function isolateGlobal(): string {
  const dir = tempDir();
  vi.stubEnv("AppData", dir);
  vi.stubEnv("XDG_CONFIG_HOME", dir);
  vi.stubEnv("HOME", dir);
  return globalPath();
}

function write(file: string, content: string): void {
  mkdirSync(path.dirname(file), { recursive: true });
  writeFileSync(file, content);
}

afterEach(() => {
  vi.unstubAllEnvs();
});

describe("load", () => {
  it("returns the defaults when nothing exists", () => {
    isolateGlobal();
    const res = load(tempDir(), "");
    expect(res.sources).toEqual([]);
    expect(res.config.model).toBe(defaults().model);
  });

  it("merges field by field", () => {
    isolateGlobal();
    const root = tempDir();
    write(path.join(root, FILE_NAME), `{"model":"opus"}`);

    const res = load(root, "");
    expect(res.config.model).toBe("opus");
    expect(res.config.subjectMaxLength).toBe(defaults().subjectMaxLength);
    expect(res.config.types).toHaveLength(defaults().types.length);
  });

  it("applies the documented precedence", () => {
    const globalFile = isolateGlobal();
    const root = tempDir();

    write(globalFile, `{"model":"from-global","language":"de","subjectMaxLength":50}`);
    write(
      path.join(root, "package.json"),
      `{"name":"x","commitron":{"model":"from-package","language":"fr"}}`,
    );
    write(path.join(root, FILE_NAME), `{"model":"from-repo"}`);

    const explicit = path.join(tempDir(), "extra.json");
    write(explicit, `{"body":"never"}`);

    const res = load(root, explicit);
    expect(res.config.model).toBe("from-repo");
    expect(res.config.language).toBe("fr");
    expect(res.config.subjectMaxLength).toBe(50);
    expect(res.config.body).toBe("never");
    expect(res.sources).toHaveLength(4);
  });

  it("rejects unknown keys and names them", () => {
    isolateGlobal();
    const root = tempDir();
    write(path.join(root, FILE_NAME), `{"modelo":"opus"}`);

    expect(() => load(root, "")).toThrow(/modelo/);
  });

  it("rejects values of the wrong type", () => {
    isolateGlobal();
    const root = tempDir();
    write(path.join(root, FILE_NAME), `{"timeoutSeconds":"120"}`);

    expect(() => load(root, "")).toThrow(/timeoutSeconds/);
  });

  it("accepts the $schema key", () => {
    isolateGlobal();
    const root = tempDir();
    write(path.join(root, FILE_NAME), `{"$schema":"${SCHEMA_URL}","model":"opus"}`);

    expect(() => load(root, "")).not.toThrow();
  });

  it("treats null instructions as none", () => {
    isolateGlobal();
    const root = tempDir();
    write(path.join(root, FILE_NAME), `{"instructions":null}`);

    expect(load(root, "").config.instructions).toBeUndefined();
  });

  it("ignores a package.json without our key", () => {
    isolateGlobal();
    const root = tempDir();
    write(path.join(root, "package.json"), `{"name":"x","scripts":{}}`);

    expect(load(root, "").sources).toEqual([]);
  });

  it("survives someone else's broken package.json", () => {
    isolateGlobal();
    const root = tempDir();
    write(path.join(root, "package.json"), `{not json`);

    expect(() => load(root, "")).not.toThrow();
  });

  it("fails when an explicit config file is missing", () => {
    isolateGlobal();
    expect(() => load(tempDir(), path.join(tempDir(), "nope.json"))).toThrow();
  });
});

File: internal/config/load_test.go
package config

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func isolateGlobal(t *testing.T) string {
	t.Helper()
	dir := t.TempDir()
	t.Setenv("AppData", dir)
	t.Setenv("XDG_CONFIG_HOME", dir)
	t.Setenv("HOME", dir)
	path, err := GlobalPath()
	if err != nil {
		t.Fatal(err)
	}
	return path
}

func write(t *testing.T, path, content string) {
	t.Helper()
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		t.Fatal(err)
	}
}

func TestLoadDefaultsWhenNothingExists(t *testing.T) {
	isolateGlobal(t)
	res, err := Load(t.TempDir(), "")
	if err != nil {
		t.Fatal(err)
	}
	if len(res.Sources) != 0 {
		t.Errorf("expected no sources, got %v", res.Sources)
	}
	if res.Config.Model != Defaults().Model {
		t.Errorf("Model = %q, want the default", res.Config.Model)
	}
}

func TestLoadMergesFieldByField(t *testing.T) {
	isolateGlobal(t)
	root := t.TempDir()

	write(t, filepath.Join(root, FileName), `{"model":"opus"}`)

	res, err := Load(root, "")
	if err != nil {
		t.Fatal(err)
	}
	if res.Config.Model != "opus" {
		t.Errorf("Model = %q, want opus", res.Config.Model)
	}
	if res.Config.SubjectMaxLength != Defaults().SubjectMaxLength {
		t.Errorf("SubjectMaxLength = %d, the file did not declare it", res.Config.SubjectMaxLength)
	}
	if len(res.Config.Types) != len(Defaults().Types) {
		t.Errorf("Types was replaced by an undeclared key")
	}
}

func TestLoadPrecedence(t *testing.T) {
	globalPath := isolateGlobal(t)
	root := t.TempDir()

	write(t, globalPath, `{"model":"from-global","language":"de","subjectMaxLength":50}`)
	write(t, filepath.Join(root, "package.json"),
		`{"name":"x","commitron":{"model":"from-package","language":"fr"}}`)
	write(t, filepath.Join(root, FileName),
		`{"model":"from-repo"}`)

	explicit := filepath.Join(t.TempDir(), "extra.json")
	write(t, explicit, `{"body":"never"}`)

	res, err := Load(root, explicit)
	if err != nil {
		t.Fatal(err)
	}
	if res.Config.Model != "from-repo" {
		t.Errorf("Model = %q, .commitron.json should win over package.json and global", res.Config.Model)
	}
	if res.Config.Language != "fr" {
		t.Errorf("Language = %q, package.json should win over global", res.Config.Language)
	}
	if res.Config.SubjectMaxLength != 50 {
		t.Errorf("SubjectMaxLength = %d, the global value should survive", res.Config.SubjectMaxLength)
	}
	if res.Config.Body != "never" {
		t.Errorf("Body = %q, --config should win", res.Config.Body)
	}
	if len(res.Sources) != 4 {
		t.Errorf("want 4 sources, got %v", res.Sources)
	}
}

func TestLoadRejectsUnknownKeys(t *testing.T) {
	isolateGlobal(t)
	root := t.TempDir()
	write(t, filepath.Join(root, FileName), `{"modelo":"opus"}`)

	if _, err := Load(root, ""); err == nil {
		t.Fatal("a misspelled key must fail loudly")
	} else if !strings.Contains(err.Error(), "modelo") {
		t.Errorf("the error should name the offending key, got: %v", err)
	}
}

func TestLoadAcceptsSchemaKey(t *testing.T) {
	isolateGlobal(t)
	root := t.TempDir()
	write(t, filepath.Join(root, FileName), `{"$schema":"`+SchemaURL+`","model":"opus"}`)

	if _, err := Load(root, ""); err != nil {
		t.Fatalf("$schema must be accepted: %v", err)
	}
}

func TestLoadIgnoresPackageJSONWithoutOurKey(t *testing.T) {
	isolateGlobal(t)
	root := t.TempDir()
	write(t, filepath.Join(root, "package.json"), `{"name":"x","scripts":{}}`)

	res, err := Load(root, "")
	if err != nil {
		t.Fatal(err)
	}
	if len(res.Sources) != 0 {
		t.Errorf("a package.json without our key is not a source, got %v", res.Sources)
	}
}

func TestLoadSurvivesBrokenPackageJSON(t *testing.T) {
	isolateGlobal(t)
	root := t.TempDir()
	write(t, filepath.Join(root, "package.json"), `{not json`)

	if _, err := Load(root, ""); err != nil {
		t.Fatalf("someone else's broken package.json is not our problem: %v", err)
	}
}

func TestLoadFailsOnMissingExplicitConfig(t *testing.T) {
	isolateGlobal(t)
	if _, err := Load(t.TempDir(), filepath.Join(t.TempDir(), "nope.json")); err == nil {
		t.Fatal("an explicitly requested config file must exist")
	}
}
